File: Python/valid-parentheses.py
class Solution:
    # @return a boolean
    def isValid(self, s):
        st = []
        d = {')':'(', ']':'[', '}':'{'}
        for i in s:
            if i in d:
                if st and d[i]==st[-1]: st.pop()
                else: return False
            else: st.append(i)
        return not st

File: C++/valid-parentheses.cpp
class Solution {
public:
    bool isValid(std::string s) {
        std::stack<char> stk;
        std::unordered_map<char, char> hashmap{{')','('}, {']', '['}, {'}', '{'}};
        for(char c : s) {
            if(hashmap.find(c)==hashmap.end()) {
                stk.push(c);
            }
            else {
                if(!stk.empty() && stk.top()==hashmap[c])
                    stk.pop();
                else
                    return false;
            }
        }
        return stk.empty();
    }
};

